### DISCUSSION TITLE
How to compute fft to store sine and cosine values and then set the higher harmonics to zero for performing Fourier smoothing in python?

### DISCUSSION THREAD

**Discussion by dikshitanavle** [pydata/xarray#4654] [Q&A]
Fourier Transform in Python can be taken by using various functions in scipy or numpy modules. I want to perform Fourier smoothing on the available data by:

Taking Fourier Transform of the available data
Setting higher harmonics to 0
And then taking its Inverse Fourier Transform
I tried this by using numpy in python as follows:

```
X=np.fft.rfft(ds)
X[3:]=0
Y=np.fft.irfft(ds)
```

I have a 4-dimensional daily dataset for air temperature which has dimensions- time, level, latitude and longitude. But I want to take Fourier transform such that the FFT coefficients: real(cosine) and imaginary(sine) are stored as two different parts. The FFT function should introduce an extra (5th) dimension for this purpose. For example, fftofds[0,0:11,6,6,6] should return the real(cosine) part at time(0:11),level(6), latitude(6), longitude(6) and , fftofds[1,0:11,6,6,6] should return the imaginary(sine) part at time(0:11),level(6), latitude(6), longitude(6).
Further the higher harmonics should be set to zero. eg: fftofds[3:]=0 and then the Inverse FFT should be performed on it. How can I achieve this using Python?

**Comment by max-sixty** [comment]
Hi @dikshitanavle — there's a lot in your question. Could you break it down — what's the xarray-specific part of the question? Could you abstract that from the FFT part of your question?

Here are some resources for asking engageable questions:
- https://matthewrocklin.com/blog/work/2018/02/28/minimal-bug-reports
- https://stackoverflow.com/help/minimal-reproducible-example

**Reply by dikshitanavle** [reply]
Thanks for your response @max-sixty

Compute sine and cosine values of fft in xarray

We are working on a netCDF4 dataset that has four dimensions, i.e., time, level, latitude, and longitude.
We are primarily using xarray for I/O and other calculations on this dataset.

The main problem we are facing is calculating sine and cosine values of Fast Fourier Transform.
We want to calculate both these values using xarray and to retain the dimensions. We tried using NumPy but didn't get the desired output.

Our code:
```
X=np.fft.rfft(ds)
X[3:]=0
Y=np.fft.irfft(ds)
```

Thank You!

**Reply by dcherian** [reply]
Please see https://xrft.readthedocs.io/en/latest/
